sphinx/source/docs/user_guide/data.rst

```diff
@@ -0,0 +1,363 @@
+will enable you access more advanced capabilites, such as streaming data,
+sharing data between plots, and filtering data.
+
+ColumnDataSource
+----------------
+
+The |ColumnDataSource| is the core of most Bokeh plots, providing the data
+that is visualized by the glyphs of the plot. With the |ColumnDataSource|,
+it is easy to share data between multiple plots and widgets, such as the
+|DataTable|. When the same |ColumnDataSource| is used to drive multiple
+renderers, selections of the data source are also shared. Thus it is possible
+to use a select tool to choose data points from one plot and have them automatically
+highlighted in a second plot (:ref:`userguide_data_linked_selection`).
+
+At the most basic level, a |ColumnDataSource| is simply a mapping between column
+names and lists of data. The |ColumnDataSource| takes a ``data`` parameter which is a dict,
+with string column names as keys and lists (or arrays) of data values as values. If one positional
+argument is passed in to the |ColumnDataSource| initializer, it will be taken as ``data``. Once the
+|ColumnDataSource| has been created, it can be passed into the ``source`` parameter of
+plotting methods which allows you to pass a column's name as a stand in for the data values:
+
+.. code-block:: python
+
+    from bokeh.plotting import figure
+    from bokeh.models import ColumnDataSource
+
+    data = {'x_values': [1, 2, 3, 4, 5],
+            'y_values': [6, 7, 2, 3, 6]}
+
+    source = ColumnDataSource(data=data)
+
+    p = figure()
+    p.circle(x='x_values', y='y_values', source=source)
+
+The ``data`` parameter can also be a Pandas ``DataFrame`` or ``GroupBy`` object.
+
+.. code-block:: python
+
+   source = ColumnDataSource(df)
+
+If a ``DataFrame`` is used, the CDS will have columns corresponding to the columns of
+the ``DataFrame``. If the ``DataFrame`` has a named index column, then CDS will also have
+a column with this name. However, if the index name (or any subname of a ``MultiIndex``)
+is ``None``, then the CDS will have a column generically named ``index`` for the index.
+
+.. code-block:: python
+
+    group = df.groupby(('colA', 'ColB'))
+    source = ColumnDataSource(group)
+
+If a ``GroupBy`` object is used, the CDS will have columns corresponding to the result of
+calling ``group.describe()``. The ``describe`` method generates columns for statistical measures
+such as ``mean`` and ``count`` for all the non-grouped orginal columns. The CDS columns are
+formed by joining original column names with the computed measure. For example, if a
+``DataFrame`` has columns ``'year'`` and ``'mpg'``. Then passing ``df.groupby('year')``
+to a CDS will result in columns such as ``'mpg_mean'``
+
+If the ``GroupBy.describe`` result has a named index column, then the CDS will also have a
+column with this name. However, if the index name (or any subname of a ``MultiIndex``) is
+``None``, then the CDS will have a column generically named ``index`` for the index.
```

Comment:
I just checked this and what you say is correct about what the name should be when you group by cly and yr. I don't know what situation would lead to the index name being None for the result of `GroupBy.describe`, so should I just take that second sentence out? I did copy this from the docstring of the CDS.